source/common/router/filter_config_discovery_impl.cc

```diff
@@ -0,0 +1,191 @@
+                            -> ThreadLocal::ThreadLocalObjectSharedPtr {
+    auto prev_config = std::dynamic_pointer_cast<ThreadLocalConfig>(previous);
+    prev_config->config_ = config;
+    return previous;
+  });
+}
+
+FilterConfigSubscription::FilterConfigSubscription(
+    const envoy::config::core::v3::ConfigSource& config_source,
+    const std::string& filter_config_name, Server::Configuration::FactoryContext& factory_context,
+    const std::string& stat_prefix, FilterConfigProviderManagerImpl& filter_config_provider_manager,
+    const std::string& subscription_id)
+    : Envoy::Config::SubscriptionBase<envoy::config::core::v3::TypedExtensionConfig>(
+          envoy::config::core::v3::ApiVersion::V3,
+          factory_context.messageValidationContext().dynamicValidationVisitor(), "name"),
+      filter_config_name_(filter_config_name), factory_context_(factory_context),
+      validator_(factory_context.messageValidationContext().dynamicValidationVisitor()),
+      init_target_(fmt::format("FilterConfigSubscription init {}", filter_config_name_),
+                   [this]() { start(); }),
+      scope_(factory_context.scope().createScope(stat_prefix + "filter_config_discovery." +
+                                                 filter_config_name_ + ".")),
+      stat_prefix_(stat_prefix), filter_config_provider_manager_(filter_config_provider_manager),
+      subscription_id_(subscription_id) {
+  const auto resource_name = getResourceName();
+  subscription_ =
+      factory_context.clusterManager().subscriptionFactory().subscriptionFromConfigSource(
+          config_source, Grpc::Common::typeUrl(resource_name), *scope_, *this, resource_decoder_);
+}
+
+void FilterConfigSubscription::start() {
+  if (!started_) {
+    started_ = true;
+    subscription_->start({filter_config_name_});
+  }
+}
+
+void FilterConfigSubscription::onConfigUpdate(
+    const std::vector<Envoy::Config::DecodedResourceRef>& resources,
+    const std::string& version_info) {
+  // Make sure to make progress in case the control plane is temporarily inconsistent.
+  init_target_.ready();
+
+  if (resources.size() != 1) {
+    throw EnvoyException(fmt::format(
+        "Unexpected number of resources in FilterConfigDS response: {}", resources.size()));
+  }
+  const auto& filter_config = dynamic_cast<const envoy::config::core::v3::TypedExtensionConfig&>(
+      resources[0].get().resource());
+  if (filter_config.name() != filter_config_name_) {
+    throw EnvoyException(fmt::format("Unexpected resource name in FilterConfigDS response: {}",
+                                     filter_config.name()));
+  }
+  auto& factory = Envoy::Config::Utility::getAndCheckFactory<
+      Server::Configuration::NamedHttpFilterConfigFactory>(filter_config);
+  // Ensure that the factory is valid in the filter chain context.
+  for (auto* provider : filter_config_providers_) {
+    provider->validateConfig(factory);
```

Comment:
I guess this is where I'm confused. I think the type of the filter should be knowable before we fetch it over xDS, hence properties such as terminal should be knowable. Does the API as it stands require us to wait until we've done an xDS fetch, retrieved the `Any`, to be able to extract type URL and know the actual filter schema?

Reply:
Yes, the type is not known ahead of time, so its validation is delayed. We can additional constraints to enforce filter types for a named resource but that's also going to be a delayed validation.

Reply:
I wonder if it makes sense to set the type ahead of time, both to perform early validation and also to provide some more typing, i.e. catch bugs. In the future, we might also have signed filter configs, which would follow this "I expect X, I will verify X when I receive the response" pattern.

Reply:
I think we should do validation outside of xDS transport. It's a good idea to add constraints on the extension type. But I don't see how to do it in the transport itself, unless xDS request is extended to send those constraints.

Reply:
Well, we can specify inside the filter chain that references xDS an expected type URL, for example. I think another reason this might matter is when we do finer grained delegation of authority, and have different control planes with different trust relationships serving up filter configs. I might say "I'll allow Super Scary control plane to configure my local rate limit filter for tenant X". Super Scary control plane shouldn't be able to deliver some Lua filter instead.

Reply:
Agree. I think it could belong to HttpFilterConfigSource. Do you want me to update the API and implement it? I am concerned about the performance of validation code since it's iterating over all providers, so passing an arbitrary function or doing proto conversion while validating seems too expensive.

Reply:
Yeah, let's fix the API. I think all we need to do is validate the terminal status of filters in the chain when we create the filter chain and then later just do type URL comparisons.